src/adapter/variables.ts

```diff
@@ -570,20 +580,60 @@ export class VariableStore {
     };
   }
 
-  private _createObjectVariable(name: string, value: RemoteObject, context?: string): Dap.Variable {
+  private async _createObjectVariable(
+    name: string,
+    value: RemoteObject,
+    context?: string,
+  ): Promise<Dap.Variable> {
     const variablesReference = this._createVariableReference(value);
     const object = value.o;
     return {
       name,
-      value:
-        (name === '__proto__' && object.description) ||
-        objectPreview.previewRemoteObject(object, context),
+      value: await this._generateVariableValueDescription(name, value, object, context),
       evaluateName: value.accessor,
       type: object.subtype || object.type,
       variablesReference,
     };
   }
 
+  private async _generateVariableValueDescription(
+    name: string,
+    value: RemoteObject,
+    object: Cdp.Runtime.RemoteObject,
+    context?: string,
+  ): Promise<string> {
+    const defaultValueDescription =
+      (name === '__proto__' && object.description) ||
+      objectPreview.previewRemoteObject(object, context);
+
+    if (this.customDescriptionGenerator) {
+      const customValueDescription = await this._cdp.Runtime.callFunctionOn({
+        objectId: object.objectId,
+        functionDeclaration: this.customDescriptionGenerator,
```

Comment:
You could reuse the infrastructure we use for [log statements](https://github.com/microsoft/vscode-js-debug/blob/master/src/adapter/breakpoints/conditions/logPoint.ts#L54-L83) for this description. This would let users write something like `JSON.stringify(x)` instead of `function (x) { return JSON.stringify(x) }`, which is more pleasant to deal with.

I also think capturing errors inline, like the logpoints do, is preferable -- I will be able to see that an error happened if I'm looking for my value, but I don't get a wall of errors in my console whenever I hit a breakpoint. It's also not always clear where the failure happened out of context, even with the name, since variable are collected for all kinds of things that may not be obvious to the user.